subsys/elfloader/elfloader.c

```diff
@@ -0,0 +1,363 @@
+					* ehdr.e_shentsize, FS_SEEK_SET);
+			fs_read(zfp, (void *)&dynstr_shdr,
+					sizeof(dynstr_shdr));
+		}
+		/* find out relocation sections */
+		else if ((shdr.sh_type == SHT_REL)
+				|| (shdr.sh_type == SHT_RELA)) {
+			rel_cnt = shdr.sh_size / sizeof(elf_rel_t);
+			for (j = 0; j < rel_cnt; j++) {
+				/* get each relocation entry */
+				fs_seek(zfp, shdr.sh_offset
+						+ j * sizeof(elf_rel_t),
+						FS_SEEK_SET);
+				fs_read(zfp, (void *)&rel, sizeof(elf_rel_t));
+
+				/* get corresponding symbol */
+				fs_seek(zfp, dynsym_shdr.sh_offset
+						+ ELF_R_SYM(rel.r_info)
+						* sizeof(elf_sym_t),
+						FS_SEEK_SET);
+				fs_read(zfp, (void *)&sym, sizeof(sym));
+
+				/* get corresponding symbol str name */
+				fs_seek(zfp, dynstr_shdr.sh_offset
+						+ sym.st_name, FS_SEEK_SET);
+				fs_read(zfp, str_name, sizeof(str_name));
+
+				/* find out symbol's real address */
+				if (sym.st_shndx == SHN_UNDEF) {
+					/* this symbol needs to be found
+					 * globally.
+					 */
+					sym_addr = (elf_addr)zmodule_find_sym(
+							KERNEL_MODULE,
+							str_name);
+				} else {
+					/* this symbol could be found locally */
+					sym_addr = module->load_start_addr
+						+ sym.st_value
+						- module->virt_start_addr;
+				}
+
+				/* doing relocation */
+				elfloader_arch_relocate_dyn(module, &rel,
+						sym_addr);
+			}
+		}
+	}
+
+	/* get all exported symbols of module.
+	 * iterate all symbols in .dynsym section and find out those symbols
+	 * with gloabl, function attributes.
+	 */
+	/* get total count of exported symbols */
+	count = 0;
+	dynsym_cnt = dynsym_shdr.sh_size / sizeof(elf_sym_t);
+	for (j = 0; j < dynsym_cnt; j++) {
+		fs_seek(zfp, dynsym_shdr.sh_offset + j * sizeof(elf_sym_t),
+				FS_SEEK_SET);
+		fs_read(zfp, (void *)&sym, sizeof(sym));
+		if ((ELF_ST_BIND(sym.st_info) == STB_GLOBAL)
+				&& (ELF_ST_TYPE(sym.st_info) == STT_FUNC)
+				&& (sym.st_shndx != SHN_UNDEF)) {
+			count++;
+		}
+	}
+	module->sym_list = (zmodule_symbol_t *)k_malloc(count
+			* sizeof(zmodule_symbol_t));
+	module->sym_cnt = count;
+
+	/* get each exported symbols's address and string name */
+	for (j = 0, count = 0; j < dynsym_cnt; j++) {
+		fs_seek(zfp, dynsym_shdr.sh_offset
+				+ j * sizeof(elf_sym_t), FS_SEEK_SET);
+		fs_read(zfp, (void *)&sym, sizeof(sym));
+		if ((ELF_ST_BIND(sym.st_info) == STB_GLOBAL)
+				&& (ELF_ST_TYPE(sym.st_info) == STT_FUNC)
+				&& (sym.st_shndx != SHN_UNDEF)) {
+
+			/* get symbol's address */
+			module->sym_list[count].addr = sym.st_value
+				- module->virt_start_addr
+				+ module->load_start_addr;
+
+			/* get symbol's string name */
+			fs_seek(zfp, dynstr_shdr.sh_offset
+					+ sym.st_name, FS_SEEK_SET);
+			fs_read(zfp, str_name, sizeof(str_name));
+			len = strlen(str_name) + 1;
+			module->sym_list[count].name = k_malloc(len);
+			memcpy((void *)module->sym_list[count].name,
+					(void *)str_name, len);
+			count++;
+		}
+	}
+}
+
+void *zmodule_load(const char *filename)
```

Comment:
FWIW: tying this API to Zephyr's, er, "immature" filesystem interface seems likely to be a pretty bad limitation.  Lots of platforms without filesystem storage are likely to want to use this.  It should be possible to read an image out of an addressible region of flash or as an offset in a SPI chip, etc...  It would be nice if we had some kind of pre-existing byte stream (i.e. "file descriptor") interface, but we don't.   This may need some kind of read/write callback or something.

Reply:
> read an image out of an addressable region of flash or as an offset in a SPI chip.

hi @andyross  do you mean the API should be like this, receive an address pointer and length as arguments?
struct zmodule *zmodule_load(void *mem, size_t length)

but how do we store the content of module object file into flash or SPI chip? when we build the module, the output it is a ELF file, how do we save this file into flash?